diff -Naur a/drivers/mtd/nand/pl35x_nand.c b/drivers/mtd/nand/pl35x_nand.c
--- a/drivers/mtd/nand/pl35x_nand.c	2016-01-25 18:34:24.065782741 -0700
+++ b/drivers/mtd/nand/pl35x_nand.c	2016-01-25 18:55:51.945806920 -0700
@@ -117,6 +117,9 @@
 	{NAND_CMD_PARAM, NAND_CMD_NONE, 1, NAND_CMD_NONE},
 	{NAND_CMD_GET_FEATURES, NAND_CMD_NONE, 1, NAND_CMD_NONE},
 	{NAND_CMD_SET_FEATURES, NAND_CMD_NONE, 1, NAND_CMD_NONE},
+	{NAND_CMD_UNLOCK1, NAND_CMD_NONE, 3, NAND_CMD_NONE},
+	{NAND_CMD_UNLOCK2, NAND_CMD_NONE, 3, NAND_CMD_NONE},
+	{NAND_CMD_LOCK, NAND_CMD_NONE, 0, NAND_CMD_NONE},
 	{NAND_CMD_NONE, NAND_CMD_NONE, 0, 0},
 	/* Add all the flash commands supported by the flash device and Linux */
 	/*
